File: src/Utils/Color.ts
export class Color {
    public static readonly WHITE = new Color(255, 255, 255).ToUint32();
    public static readonly BLACK = new Color(0, 0, 0).ToUint32();
    public static readonly GRAY = new Color(128, 128, 128).ToUint32();
    public static readonly RED = new Color(255, 0, 0).ToUint32();
    public static readonly GREEN = new Color(0, 255, 0).ToUint32();
    public static readonly BLUE = new Color(0, 0, 255).ToUint32();
    public static readonly YELLOW = new Color(255, 255, 0).ToUint32();
    public static readonly CYAN = new Color(0, 255, 255).ToUint32();
    public static readonly MAGENTA = new Color(255, 0, 255).ToUint32();
    public static readonly ORANGE = new Color(255, 165, 0).ToUint32();
    public static readonly PURPLE = new Color(128, 0, 128).ToUint32();

    public r: number;
    public g: number;
    public b: number;
    public a: number;

    constructor(r: number, g: number, b: number, a: number = 255) {
        this.r = r;
        this.g = g;
        this.b = b;
        this.a = a;
    }

    public ToUint32() {
        return (this.a << 24) | (this.b << 16) | (this.g << 8) | this.r;
    }

    public static FromUint32(uint32: number) {
        return new Color(
            uint32 & 0xFF,
            (uint32 >> 8) & 0xFF,
            (uint32 >> 16) & 0xFF,
            (uint32 >> 24) & 0xFF
        );
    }
}

File: C++/src/color.cpp
#include "color.h"

// 预制颜色的定义
const Color Color::White = Color(255, 255, 255);
const Color Color::Black = Color(0, 0, 0);
const Color Color::Red = Color(255, 0, 0);
const Color Color::Green = Color(0, 255, 0);
const Color Color::Blue = Color(0, 0, 255);
const Color Color::Cyan = Color(0, 255, 255);
const Color Color::Yellow = Color(255, 255, 0);
const Color Color::Magenta = Color(255, 0, 255);
const Color Color::Gray = Color(128, 128, 128);

Color Color::operator*(float t) const
{
    // 对每个通道进行乘法运算，并确保结果不超过255
    return {static_cast<uint8_t>(r * t),
            static_cast<uint8_t>(g * t),
            static_cast<uint8_t>(b * t),
            static_cast<uint8_t>(a * t)};
}

Color Color::operator+(const Color &c) const
{
    //TODO:这个算法明显不对，两个透明度255的颜色相加，透明度反而变成254了
    //  对每个通道进行相加，并确保结果不超过255
    return {static_cast<uint8_t>(r + c.r),
            static_cast<uint8_t>(g + c.g),
            static_cast<uint8_t>(b + c.b),
            static_cast<uint8_t>(a + c.a)};
}

Color Color::interpolateColor(const Color &c1, const Color &c2, float t)
{
    uint8_t r = static_cast<uint8_t>(c1.r + t * (c2.r - c1.r));
    uint8_t g = static_cast<uint8_t>(c1.g + t * (c2.g - c1.g));
    uint8_t b = static_cast<uint8_t>(c1.b + t * (c2.b - c1.b));
    uint8_t a = static_cast<uint8_t>(c1.a + t * (c2.a - c1.a));
    return Color(r, g, b, a);
}